src/dune/workspace.ml

```diff
@@ -165,29 +166,38 @@ module Context = struct
       let open Dyn.Encoder in
       record
         [ ("base", Common.to_dyn base)
-        ; ("switch", Context_name.to_dyn switch)
+        ; ("switch", string switch)
         ; ("root", option string root)
         ; ("merlin", bool merlin)
         ]
 
     let equal { base; switch; root; merlin } t =
       Common.equal base t.base
-      && Context_name.equal switch t.switch
+      && String.equal switch t.switch
       && Option.equal String.equal root t.root
       && Bool.equal merlin t.merlin
 
     let t ~profile ~x =
-      let+ switch = field "switch" Context_name.decode
+      let+ loc_switch, switch = field "switch" (located string)
       and+ name = field_o "name" Context_name.decode
       and+ root = field_o "root" string
       and+ merlin = field_b "merlin"
       and+ base = Common.t ~profile in
-      let default =
-        (* TODO this needs proper error handling with locations *)
-        let name = Context_name.to_string switch ^ Common.fdo_suffix base in
-        Context_name.parse_string_exn (Loc.none, name)
+      let name =
+        match name with
+        | Some s -> s
+        | None -> (
+          let name = Filename.basename switch ^ Common.fdo_suffix base in
+          match Context_name.of_string_opt name with
```

Comment:
Not specific to this PR, but I remember that once we discussed the naming of these functions and agreed on:

- `of_string : string -> t` for places where we are just converting between data structures and don't expect errors
- `parse_string : string -> (t, ...) result` for places where we are parsing data coming from the outside world (here the `...` could just be `unit`)

Reply:
Indeed we have. Here's how interface looks like:

```
  val of_string : string -> t

  val parse_string_exn : Loc.t * string -> t
```

So we all need to do is to remove the `_exn` suffix from parsing.